portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.sps.data.LoginInfo;
+import com.google.gson.Gson;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that grabs a user's login status and dispatches
+ *  to DataServlet to display its content according to the status.
+ */
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+    private final Gson gson = new Gson();
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        response.setContentType("application/json;");
+        PrintWriter out = response.getWriter();
+        boolean loggedIn;
+        String url;
+        String userEmail;
+
+        UserService credentials = UserServiceFactory.getUserService();
+        if(credentials.isUserLoggedIn()) {
+            loggedIn = true;
+            url = credentials.createLogoutURL("/homepage.html");
+            userEmail = credentials.getCurrentUser().getEmail();
+        } else {
+            loggedIn = false;
+            url = credentials.createLoginURL("/homepage.html");
+            userEmail = "";
+        }
+
+        LoginInfo userLogin = new LoginInfo(loggedIn, url, userEmail);
+        request.setAttribute("userLogin", userLogin);
+        RequestDispatcher dispatch = request.getRequestDispatcher("/data");
+        dispatch.forward(request, response);
```

Comment:
What would you do if you had user data other than comments? For example, what if you have a page that lets users upload and view images posted by each other, separately from comments? You wouldn't want to have to send back all information for both page types for every request; as your application got larger with more pages, you would keep having to send back more and more data from `/login` even though each page only displays one thing.

Handling things this way is also confusing: if you want (in the JavaScript) to check whether the user is logged in and get a login/logout URL for them, you use `/login`. But if you want to fetch their comment data, you *also* use `/login`.

It would be a lot clearer about what was going on if you keep these actions separate. Keep things simple! `/login` should just handle getting login/logout URLs. The JS should use `/data` (or rename it to `/comments`) to fetch comment data. Just call the UserService to check if the user is logged in before retrieving any sensitive data in any other servlet, and if not use `response.setStatus(HttpServletResponse.SC_FORBIDDEN);` and send an empty json response.

If you really want to get into how this sort of check might be done more consistently in practice (so every servlet doesn't need to do its own check), you could look into `Filter`s, but for this project it may just be easier to put user service checks in any data-access endpoints.

Reply:
Yes I agree it was confusing to create and to understand. I wanted the servlets to communicate with eachother and have their own responsibilities while still passing necessary information back to the client side. I wanted to avoid using UserService since I wanted the login credentials to be solely handled by the LoginServlet, but it turned messy. In hindsight, UserService keeps the logged in information and would be accessible wherever it is called in servlets, but I wanted to try to incorporate it in LoginServlet only. Trying to fully-separate tasks for servlets to handle on their own seemed to add an additional layer of complexity and confusion.

Reply:
I think part of what makes some of these things seem more complicated is that you're using a very barebones server framework and no JS framework, so setting up different parts of your page sending requests to different servlets probably seems pretty complicated. Normally with a JS framework like React, you'd have different parts of the page or a background data management system to fetch relevant data for different parts of the page, and render it into HTML.

I think even without this kind of framework you can still us this principle. For example; when you build out HTML using document.createElement calls, you don't actually have to attach the HTML you're building to any particular part of the page within that function. Instead you could create an HTML subtree and return it, and let the caller decide where to attach it. This would allow you to, for example, create a single function that builds the comment display area, but which isn't tied to any particular page. Then to use it on a particular page, you just call it to create the correct tree, then appendChild it onto the correct element in the calling function.

Once you start creating JS functions like this, I think you'll find that having servlets separate will be a bit easier. You can use a setup like:

```javascript
checkLoggedIn().then((loginInfo) => {
  if (loginInfo.isLoggedIn) {
    return loadCommentData().then((commentData) => renderComments(commentData));
  } else {
    return renderLoginForm(loginInfo);
  }
}).then((commentAreaOrLoginForm) => {
  var commentArea = document.findElementById("comment-area");
  commentArea.innerHTML = "";
  commentArea.appendChild(commentAreaOrLoginForm);
});
```